File: Leetcode/C++/53. Maximum Subarray/solution.cpp
// https://leetcode.com/problems/maximum-subarray/description/
// Time complexity: O(N)

#include <algorithm>
#include <vector>

using namespace std;

class Solution
{
public:
    int maxSubArray(vector<int> &nums)
    {
        int maxSum = INT_MIN;
        int total = 0;

        for (auto num : nums)
        {
            total += num;

            if (maxSum < total)
            {
                maxSum = total;
            }
            if (total < 0)
            {
                total = 0;
            }
        }

        return maxSum;
    }
};

class Solution // Kadane`s algorithm
{
public:
    int maxSubArray(vector<int> &nums)
    {
        int maxSum = INT_MIN, curSum = 0;
        for (auto num : nums)
        {
            curSum = max(curSum + num, num);
            maxSum = max(maxSum, curSum);
        }
        return maxSum;
    }
};

File: Leetcode/Python/53. Maximum Subarray/solution.py
"""
https://leetcode.com/problems/maximum-subarray/
Time complexity: O(N)
"""


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        for idx in range(1, len(nums)):
            # Add to the current value only when the previous sums are positive.
            if nums[idx - 1] > 0:
                nums[idx] += nums[idx - 1]
        return max(nums)


class Solution: # Kadane`s algorithm
    def maxSubArray(self, nums: List[int]) -> int:
        best_sum = -float("inf")
        current_sum = 0
        for num in nums:
            current_sum = max(num, current_sum + num)
            best_sum = max(best_sum, current_sum)
        
        return best_sum